src/amo/components/SearchForm/index.js

```diff
@@ -0,0 +1,211 @@
+  validAddonTypes,
+} from 'core/constants';
+import translate from 'core/i18n/translate';
+import { convertFiltersToQueryParams } from 'core/searchUtils';
+import SearchInput from 'ui/components/SearchInput';
+import LoadingText from 'ui/components/LoadingText';
+import { withErrorHandler } from 'core/errorHandler';
+import {
+  autocompleteAbort,
+  autocompleteStart,
+} from 'core/reducers/autocomplete';
+import Suggestion from 'amo/components/SearchForm/Suggestion';
+
+import 'core/css/inc/lib.scss';
+import './style.scss';
+
+const LOADING_SUGGESTION_KEY = 'suggestion-is-loading';
+
+export class SearchFormBase extends React.Component {
+  static propTypes = {
+    addonType: PropTypes.string,
+    api: PropTypes.object.isRequired,
+    i18n: PropTypes.object.isRequired,
+    pathname: PropTypes.string.isRequired,
+    query: PropTypes.string.isRequired,
+    dispatch: PropTypes.func.isRequired,
+    errorHandler: PropTypes.object.isRequired,
+    suggestions: PropTypes.arrayOf(PropTypes.string).isRequired,
+    loadingSuggestions: PropTypes.bool.isRequired,
+  }
+
+  static contextTypes = {
+    router: PropTypes.object,
+  }
+
+  constructor(props: Object) {
+    super(props);
+
+    this.state = {
+      search: props.query || '',
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { query } = nextProps;
+
+    if (this.props.query !== query) {
+      this.setState({ search: query || '' });
+    }
+  }
+
+  getSuggestions() {
+    if (this.props.loadingSuggestions) {
+      return Array(10).fill(LOADING_SUGGESTION_KEY);
+    }
+
+    return this.props.suggestions;
+  }
+
+  goToSearch(query) {
+    const { addonType, api, pathname } = this.props;
+    const filters = { query };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    this.context.router.push({
+      pathname: `/${api.lang}/${api.clientApp}${pathname}`,
+      query: convertFiltersToQueryParams(filters),
+    });
+  }
+
+  handleSearch = (e) => {
+    e.preventDefault();
+    this.searchInput.blur();
+    this.goToSearch(this.searchInput.value);
+  }
+
+  handleSearchChange = (e) => {
+    this.setState({ search: e.target.value });
+  }
+
+  handleSuggestionsFetchRequested = ({ value }) => {
+    if (!value) {
+      return;
+    }
+
+    const { addonType, dispatch, errorHandler } = this.props;
+    const filters = { query: value };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    dispatch(autocompleteStart({
+      errorHandlerId: errorHandler.id,
+      filters,
+    }));
+  }
+
+  handleSuggestionsClearRequested = () => {
+    this.props.dispatch(autocompleteAbort());
+  }
+
+  handleSuggestionSelected = (e, { suggestionValue }) => {
+    this.setState({ search: suggestionValue }, () => {
+      this.goToSearch(suggestionValue);
+    });
+  }
+
+  renderSuggestion = (suggestion) => {
+    const { search } = this.state;
+
+    if (this.props.loadingSuggestions && suggestion === LOADING_SUGGESTION_KEY) {
```

Comment:
Is `suggestion === LOADING_SUGGESTION_KEY` needed when the component is already marked as loading (and thus wanting to render `<LoadingText>`?